### PR TITLE
service/dsp: Clean up global state

### PR DESCRIPTION
**Pull request by NarcolepticK**
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3996)

### CODE REVIEW

File: src/audio_core/dsp_interface.h [hunk 1]
```diff
@@ -60,6 +66,9 @@ class DspInterface {
     /// Returns a reference to the array backing DSP memory
     virtual std::array<u8, Memory::DSP_RAM_SIZE>& GetDspMemory() = 0;
 
+    /// Sets the dsp class that we trigger interrupts for
+    virtual void SetDspToInterrupt(std::shared_ptr<Service::DSP::DSP_DSP> dsp) = 0;
```
Comment (MerryMage) [inline src/audio_core/dsp_interface.h]: Consider a std::weak_ptr instead. Consider also renaming this to ‘SetServiceToInterrupt’.

File: src/audio_core/hle/hle.cpp [hunk 2]
```diff
@@ -60,6 +64,8 @@ struct DspHle::Impl final {
 
     DspHle& parent;
     CoreTiming::EventType* tick_event;
+
+    std::shared_ptr<DSP_DSP> dsp_dsp;
```
Comment (MerryMage) [inline src/audio_core/hle/hle.cpp]: std::weak_ptr as well here please.

File: src/audio_core/hle/hle.cpp [hunk 3]
```diff
@@ -231,7 +242,8 @@ void DspHle::Impl::AudioPipeWriteStructAddresses() {
         WriteU16(DspPipe::Audio, addr);
     }
     // Signal that we have data on this pipe.
-    Service::DSP::SignalPipeInterrupt(DspPipe::Audio);
+    if (dsp_dsp)
```
Comment (MerryMage) [inline src/audio_core/hle/hle.cpp]: The `lock` call should occur here, when we want to use the object, in the `if` statement, like this:

```
if (auto service = dsp_dsp.lock()) {
   service->SignalInterrupt(InterruptType::Pipe, DspPipe::Audio);
}
```

Notice that we do not hold on to the shared_ptr from `lock` for longer than necessary.

File: src/audio_core/hle/hle.cpp (hunk 2 above)
Comment (NarcolepticK) [reply src/audio_core/hle/hle.cpp]: Right, sorry. Thank you for the added clarification. Will fix it up.

File: src/audio_core/hle/hle.cpp (hunk 3 above)
Comment (NarcolepticK) [reply src/audio_core/hle/hle.cpp]: Thank you, this helps. I'm still refreshing on the ownership semantics; I started back when c++ was in its infancy, and the past decade of changes are a bit broken. It is more of a refresh/review, than a learn from scratch.

**Pull request merged**